### PR TITLE
Feat: setup audio component

### PR DESCRIPTION
**Pull request by JujieX**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
setup audio feature
UML API design
image

### Other information:
Playground: https://github.com/galacean/galacean.github.io/pull/752

## Summary by CodeRabbit

- **New Features**
	- Introduced support for audio asset types, allowing for the management of audio clips.
	- Added classes for handling audio playback, including `AudioClip`, `AudioSource`, and `AudioManager`.
	- Implemented loading functionality for audio assets through the `AudioLoader` class.

- **Bug Fixes**
	- Enhanced error handling in audio loading and playback processes.

- **Tests**
	- Added unit tests for the `AudioSource` component to verify loading and playback functionality.

- **Documentation**
	- Updated module exports to include new audio-related classes and functionalities.

### CODE REVIEW

File: packages/loader/src/AudioLoader.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { resourceLoader, Loader, AssetPromise, AssetType, LoadItem, AudioManager } from "@galacean/engine-core";
+
+@resourceLoader(AssetType.Audio, ["mp3", "ogg", "wav"], false)
+class AudioLoader extends Loader<AudioBuffer> {
+  load(item: LoadItem): AssetPromise<AudioBuffer> {
+    return new AssetPromise((resolve) => {
+      this.request<ArrayBuffer>(item.url, { type: "arraybuffer" }).then((arrayBuffer) => {
+        AudioManager.context.decodeAudioData(arrayBuffer).then((result) => {
+          resolve(result);
```
Comment (cptbtptpbcptdtptp) [inline packages/loader/src/AudioLoader.ts:9]: reject

File: packages/core/src/audio/AudioSource.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    super(entity);
+    this._onPlayEnd = this._onPlayEnd.bind(this);
+
+    this._context = AudioManager.context;
+    this._gainNode = AudioManager.context.createGain();
+    this._gainNode.connect(AudioManager.listener);
+  }
+
+  /** play the sound from the very beginning */
+  public play() {
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/audio/AudioSource.ts:130]: delete public

Comment (cptbtptpbcptdtptp) [reply packages/core/src/audio/AudioSource.ts:130]: : void

File: packages/core/src/audio/AudioSource.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    return this._endTime;
+  }
+
+  set endTime(value: number) {
+    this._endTime = value;
+    this._duration = this._endTime - this._startTime;
+  }
+
+  /** Current playback progress, in seconds */
+  get position(): number {
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/audio/AudioSource.ts:110]: Position?

File: packages/core/src/audio/AudioSource.ts (hunk 2 above)
Comment (JujieX) [reply packages/core/src/audio/AudioSource.ts]: done

File: packages/core/src/audio/AudioSource.ts (hunk 3 above)
Comment (JujieX) [reply packages/core/src/audio/AudioSource.ts]: changed to "time"

File: packages/loader/src/AudioLoader.ts (hunk 1 above)
Comment (JujieX) [reply packages/loader/src/AudioLoader.ts]: done

File: packages/core/src/audio/AudioSource.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
+import { Component } from "../Component";
+import { Entity } from "../Entity";
+import { AudioClip } from "./AudioClip";
+import { AudioManager } from "./AudioManager";
+import { deepClone, ignoreClone } from "../clone/CloneManager";
+
+/**
+ * Audio Source Component
+ */
+export class AudioSource extends Component {
```
Comment (singlecoder) [inline packages/core/src/audio/AudioSource.ts:10]: AudioSource like a resource, not component

File: packages/core/src/audio/AudioSource.ts [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        : this.engine.time.elapsedTime - this._absoluteStartTime + this.startTime;
+    }
+    return 0;
+  }
+
+  /**
+   * Plays the clip.
+   */
+  play(): void {
+    if (!this._clip || !this.clip.duration || this.isPlaying) return;
```
Comment (singlecoder) [inline packages/core/src/audio/AudioSource.ts:158]: this.clip.duration why not this._clip.duration?

File: packages/core/src/audio/AudioSource.ts [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    if (this.startTime > this._clip.duration || this.startTime < 0) return;
+    if (this._duration && this._duration < 0) return;
+
+    this._pausedTime = null;
```
Comment (singlecoder) [inline packages/core/src/audio/AudioSource.ts:162]: _pausedTime type is number

File: packages/core/src/audio/AudioSource.ts [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+  private _sourceNode: AudioBufferSourceNode;
+
+  @deepClone
+  private _startTime: number = 0;
+  @deepClone
+  private _pausedTime: number = null;
+  @deepClone
+  private _endTime: number = null;
+  @deepClone
+  private _duration: number = null;
```
Comment (singlecoder) [inline packages/core/src/audio/AudioSource.ts:29]: number init null is not good

File: packages/core/src/audio/AudioClip.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+export class AudioClip extends ReferResource {
+  /** the name of clip */
+  name: string;
+
+  private _audioBuffer: AudioBuffer;
+
+  /**
+   * the number of discrete audio channels
+   */
+  get channels(): Readonly<number> {
```
Comment (GuoLei1990) [inline packages/core/src/audio/AudioClip.ts:16]: `get channels(): number` is OK

File: packages/core/src/audio/AudioClip.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+ * Audio Clip
+ */
+export class AudioClip extends ReferResource {
+  /** the name of clip */
+  name: string;
+
+  private _audioBuffer: AudioBuffer;
+
+  /**
+   * the number of discrete audio channels
```
Comment (GuoLei1990) [inline packages/core/src/audio/AudioClip.ts:14]: Comments need to add a period，param comments do not need

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]
Unit test must add!

File: packages/core/src/audio/AudioClip.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+   * the duration, in seconds
+   */
+  get duration(): Readonly<number> {
+    return this._audioBuffer.duration;
+  }
+
+  /**
+   * get the clip's audio buffer
+   */
+  getData(): AudioBuffer {
```
Comment (GuoLei1990) [inline packages/core/src/audio/AudioClip.ts:37]: Avoid web native interfaces as much as possible

File: packages/core/src/audio/index.ts [hunk 11]
```diff
@@ -0,0 +1,4 @@
+export { AudioManager } from "./AudioManager";
```
Comment (GuoLei1990) [inline packages/core/src/audio/index.ts:1]: AudioManager is internal

File: packages/core/src/audio/AudioListener.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+ */
+export class AudioListener extends Component {
+  /**
+   * @internal
+   */
+  constructor(entity: Entity) {
+    super(entity);
+    const gain = AudioManager.context.createGain();
+    gain.connect(AudioManager.context.destination);
+    AudioManager.listener = gain;
```
Comment (GuoLei1990) [inline packages/core/src/audio/AudioListener.ts:17]: Only allowed one listener？

File: packages/core/src/audio/AudioListener.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+ * Can only have one in a scene.
+ */
+export class AudioListener extends Component {
+  /**
+   * @internal
+   */
+  constructor(entity: Entity) {
+    super(entity);
+    const gain = AudioManager.context.createGain();
+    gain.connect(AudioManager.context.destination);
```
Comment (GuoLei1990) [inline packages/core/src/audio/AudioListener.ts:16]: should be encapsulated

File: packages/core/src/audio/AudioManager.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+  /**
+   * Audio context
+   */
+  static get context(): AudioContext {
+    if (!AudioManager._context) {
+      AudioManager._context = new window.AudioContext();
+    }
+    if (AudioManager._context.state !== "running") {
+      window.document.addEventListener("pointerdown", AudioManager._unlock, true);
```
Comment (GuoLei1990) [inline packages/core/src/audio/AudioManager.ts:21]: Should do this when call `context` every time?

File: packages/loader/src/AudioLoader.ts [hunk 15]
```diff
@@ -0,0 +1,19 @@
+import { resourceLoader, Loader, AssetPromise, AssetType, LoadItem, AudioManager } from "@galacean/engine-core";
+
+@resourceLoader(AssetType.Audio, ["mp3", "ogg", "wav"], false)
+class AudioLoader extends Loader<AudioBuffer> {
+  load(item: LoadItem): AssetPromise<AudioBuffer> {
+    return new AssetPromise((resolve, reject) => {
+      this.request<ArrayBuffer>(item.url, { type: "arraybuffer" }).then((arrayBuffer) => {
+        AudioManager.context
+          .decodeAudioData(arrayBuffer)
```
Comment (GuoLei1990) [inline packages/loader/src/AudioLoader.ts:9]: should do some encapsulation

File: packages/core/src/audio/AudioSource.ts [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+  /**
+   * Unpause the paused playback of this AudioSource.
+   */
+  unPause(): void {
+    if (!this.isPlaying && this._pausedTime) {
+      const duration = this.endTime ? this.endTime - this._pausedTime : null;
+      this._play(this._pausedTime, duration);
+    }
+  }
```
Comment (GuoLei1990) [inline packages/core/src/audio/AudioSource.ts:198]: Very strange API!

File: packages/core/src/audio/AudioSource.ts [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+
+  /**
+   * Playback position in seconds.
+   */
+  get time(): number {
+    if (this.isPlaying) {
+      return this._pausedTime
+        ? this.engine.time.elapsedTime - this._absoluteStartTime + this._pausedTime
+        : this.engine.time.elapsedTime - this._absoluteStartTime + this.startTime;
```
Comment (GuoLei1990) [inline packages/core/src/audio/AudioSource.ts:149]: looks a bit complicated

File: packages/core/src/audio/AudioSource.ts [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+  }
+
+  /**
+   * The playback speed of the audio source, 1.0 is normal playback speed.
+   */
+  get playbackRate(): number {
+    return this._playbackRate;
+  }
+
+  set playbackRate(value: number) {
```
[the remark this answers is not in the document]
Comment (JujieX) [reply packages/core/src/audio/AudioSource.ts:91]: use rate

File: packages/core/src/audio/AudioSource.ts [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+    if (this._duration && this._duration < 0) return;
+
+    this._pausedTime = null;
+    this._play(this.startTime, this._duration);
+  }
+
+  /**
+   * Stops playing the clip.
+   */
+  stop(): void {
```
[the remark this answers is not in the document]
Comment (JujieX) [reply packages/core/src/audio/AudioSource.ts:158]: yes

[the remark this answers is not in the document]
Comment (JujieX) [reply packages/core/src/audio/AudioSource.ts:158]: now only has play stop and pause, removed unpause

File: packages/loader/src/AudioLoader.ts (hunk 15 above)
Comment (JujieX) [reply packages/loader/src/AudioLoader.ts:17]: now it loads onto audioClip, not audio buffer

File: packages/core/src/audio/AudioClip.ts (hunk 10 above)
Comment (JujieX) [reply packages/core/src/audio/AudioClip.ts:39]: I think getData and setData is a must

File: packages/core/src/audio/index.ts (hunk 11 above)
Comment (JujieX) [reply packages/core/src/audio/index.ts]: no export now

**Comment by GuoLei1990** [comment]
Migrated to: https://github.com/galacean/engine/pull/2449

**Pull request closed without merging**